### PR TITLE
Add B027: empty method in abstract base class with no abstract decorator

### LINKED ISSUE #277: Multi-inheritance B024
Hi, 

The latest release introduced a new error code (B024) which checks that an abc defines abstract methods/properties. I have an abc, `A`, which has a `__call__` abstract method. 

    class A(metaclass=abc.ABCMeta):
  
      def __eq__(self, other: Any) -> bool:
          ...
  
      def __repr__(self) -> str:
          ...
  
      @abc.abstractmethod
      def __call__(self, ...) -> ...:
          raise NotImplementedError`

I also have a different abc, `RandomA` which shares `A`'s methods (such as `eq`) and abstract `__call__`, but also has a random number generator as an instance variable

    class RandomA(A, metaclass=abc.ABCMeta): 
  
        def __init__(self, rng: Optional[np.random.Generator]) -> None:
            self.rng = rng if rng is not None else np.random.default_rng()

This way I can create child classes which have deterministic or random behaviour. 

`RandomA` now violates B024 as it doesn't define abstract methods, even though it should inherit `A`'s `__call__` method right?

### LINKED ISSUE #278: False positive B024
Hey! A new release (with https://github.com/PyCQA/flake8-bugbear/issues/273 included) raises an error

> Injectable is an abstract base class, but it has no abstract methods. Remember to use @ abstractmethod, @abstractclassmethod and/or @abstractproperty decorators.

I think that's fine to use ABC without abstractmethods/etc., because that's not the only thing ABCs do. They also allow to register other classes as being subclasses of the base class - see example at https://docs.python.org/3/library/abc.html#abc.ABCMeta.register. 

This is how ABC is used in our project (see https://github.com/scrapinghub/web-poet/blob/944dcd75f6c56988d42c56c56adb1b3dcafa2dbd/web_poet/pages.py#L11); a new flake8-bugbear release flags it as an error.

### LINKED ISSUE #280: False positive B024 - custom metaclass derived from ABCMeta treated as abstract base class
Another false positive with the recent B024 is when implementing custom metaclass extended from ABCMeta - they shouldn't be classified as abstract base classes at all:

```
class MyMeta(abc.ABCMeta):  # THIS GETS FALSE POSITIVE B024 BUT IS PERFECTLY LEGIT
    def __new__(mcs, name, bases, namespace):
        #  whatever custom metaclass magic here
        return super().__new__(mcs, name, bases, namespace)

class MyAbstract(metaclass=MyMeta):
    @abstractmethod
    def foobar(self):
        pass
```

```
B024 MyMeta is an abstract base class, but it has no abstract methods. Remember to use @abstractmethod, @abstractclassmethod and/or @abstractproperty decorators.
class MyMeta(abc.ABCMeta):
```

-> `MyMeta` is actually not an abstract base class (instead, it is a custom metaclass for creating abstract base classes) so this is false positive.

### PR DESCRIPTION
**Pull request by jakkdl**
EDIT: I read through the issues and realized that B024 needs changes, so made this a draft for now.

Implements the second suggested check in #273 

There seems to be some difference in style on ellipsis vs pass, so one might want to restrict the check to only be if the body is docstring and ellipsis.

When reading the docs on abc I also found out that the abstract decorators inside it other than `@abstractmethod` are [deprecated](https://docs.python.org/3/library/abc.html?highlight=abc#abc.abstractclassmethod) so I modified the b024 message to not suggest deprecated decorators.
Modified the b024 test to not trigger b027

Fixes #277, fixes #278 (though see #290), and fixes #280.

### CODE REVIEW

**Comment by jakkdl** [comment]
Updated PR to also fix #277 #280 and the original problem in #278 (but not https://github.com/PyCQA/flake8-bugbear/issues/278#issuecomment-1228146085).
ready for review and everything

File: bugbear.py [hunk 1]
```diff
@@ -632,16 +634,57 @@ def is_abstract_decorator(expr):
...
+                    isinstance(stmt, ast.Expr)
+                    and (
+                        # python<3.8
+                        isinstance(stmt.value, (ast.Ellipsis, ast.Str))
+                        or (
+                            # python3.8+
+                            isinstance(stmt.value, ast.Constant)
+                            and (
+                                stmt.value.value is Ellipsis
+                                or isinstance(stmt.value.value, str)
```
Comment (Zac-HD) [inline bugbear.py:651]: I'd pull this out as a a new helper, `is_str_or_ellipsis(node)`, and expect the code to get more readable.

**Review by Zac-HD — COMMENTED** [review]
Two comments below; otherwise looks good!

File: README.rst [hunk 2]
```diff
@@ -305,6 +308,11 @@ Change Log
 ~~~~~~~~~~
 
 * add B025: find duplicate except clauses (#284)
+=======
+Future
+~~~~~~~~~
+* Add B025: Empty method in abstract base class with no abstract decorator.
+>>>>>>> 0937a2d (Adding B025: empty method in abstract base class with no abstract decorator)
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply README.rst]: oopsies.
Time to add `check-merge-conflict` with `--assume-in-merge` to my pre-commit configs going forward

**Comment by jakkdl** [comment]
fixed both comments :+1:

**Comment by jakkdl** [comment]
Fixing #293 as well now.
Function now runs into C901, but I'm not seeing any great way of getting out of that other than moving out some of the helper functions - but that seems mediocre when they're only used in this check. Splitting the B024 and B027 into two separate functions doesn't seem great either.

**Comment by Zac-HD** [comment]
Thanks again @jakkdl!

**Pull request merged**